File: web/packages/dtx-core/src/scoring/judgment.ts
/**
 * Judgment windows in milliseconds (absolute delta between hit time and
 * target chip time). Ported from STHitRanges.cs:43-49 default values.
 *
 * A hit qualifies for the tightest window it fits; anything beyond POOR is MISS.
 */
export const HIT_RANGES_MS = {
  PERFECT: 34,
  GREAT: 67,
  GOOD: 84,
  POOR: 117,
} as const;

export const Judgment = {
  PERFECT: 'PERFECT',
  GREAT: 'GREAT',
  GOOD: 'GOOD',
  POOR: 'POOR',
  MISS: 'MISS',
} as const;

export type JudgmentKind = (typeof Judgment)[keyof typeof Judgment];

export function classifyDeltaMs(deltaMs: number): JudgmentKind {
  const d = Math.abs(deltaMs);
  if (d <= HIT_RANGES_MS.PERFECT) return Judgment.PERFECT;
  if (d <= HIT_RANGES_MS.GREAT) return Judgment.GREAT;
  if (d <= HIT_RANGES_MS.GOOD) return Judgment.GOOD;
  if (d <= HIT_RANGES_MS.POOR) return Judgment.POOR;
  return Judgment.MISS;
}

File: native/quest-spatial/app/src/main/kotlin/com/dtxmania/quest/dtxcore/scoring/Judgment.kt
package com.dtxmania.quest.dtxcore.scoring

import kotlin.math.abs

/**
 * Judgment windows in milliseconds (absolute delta between hit time and
 * target chip time). Ported from STHitRanges.cs:43-49 default values.
 *
 * A hit qualifies for the tightest window it fits; anything beyond POOR
 * is MISS.
 */
object HitRangesMs {
    const val PERFECT = 34
    const val GREAT = 67
    const val GOOD = 84
    const val POOR = 117
}

enum class Judgment { PERFECT, GREAT, GOOD, POOR, MISS }

fun classifyDeltaMs(deltaMs: Double): Judgment {
    val d = abs(deltaMs)
    return when {
        d <= HitRangesMs.PERFECT -> Judgment.PERFECT
        d <= HitRangesMs.GREAT -> Judgment.GREAT
        d <= HitRangesMs.GOOD -> Judgment.GOOD
        d <= HitRangesMs.POOR -> Judgment.POOR
        else -> Judgment.MISS
    }
}
